docs/design/backups.md

```diff
@@ -0,0 +1,83 @@
+# Backup & Restore
+
+In particularly disastrous situations, we would like to be able to revert the etcd cluster to a previously known healthy state.
+This document lays out the design of how we want the healthy state to be captured, securely stored & safely restored to operator-managed etcd clusters.
+
+## Backup Resource
+
+We will introduce a new custom resource: `EtcdBackup`.
+
+```yaml
+apiVersion: etcd.improbable.io/v1alpha1
+kind: EtcdBackup
+metadata:
+  name: my-cluster-backup
+  namespace: default
+spec:
+  type: snapshot
+  destination:
+    gcsBucket:
+      bucketName: my-cluster-backup-bucket
+      creds:
+        secretRef: 
+          name: my-cluster-backup-bucket-credentials
+          namespace: default
+```
+
+This resource specifies that we desire a backup to be immediately taken as a [snapshot](https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/recovery.md#snapshotting-the-keyspace) from the etcd API.
+This snapshot will be saved to a GCS bucket, accessed with credentials read from a given secret.
+The file would be named with a timestamp when the backup was started, for uniqueness.
+Backups will be taken by the controller, rather than a Job or Pod acting on behalf of the controller.
+
+Backup types could include
+  * `snapshot`: a `snapshot.db` file taken from the etcd API
+  * `keys`: a zip file containing all key-values as files
+  * `keys-json` a json file containing all key-values as a JSON file
+
+Only `snapshot` will be initially implemented.
+
+Destinations culd include
+  * `socket`: data would be sent to a TCP socket
+  * `gcsBucket`: pushed to a a Google Cloud Storage bucket
+  * `s3Bucket`: similar to gcsBucket, but for AWS.
+  * ... other hosted blob storage services
+
+Only `gcsBucket` will be initially implemented.
```

Comment:
Might be worth mentioning or linking to Velero's approach of having separate CRDs in their v1 API, and note that we may consider utilising them in future and opted to not include a similar construct for now to make initial implementation easier?

Reply:
I think a seperate CRD for a `BackupDestination` is a cleaner idea. EFVn if we only support GCS at start.

Reply:
from my point of view, it seems Velero is building a different thing to us, and has architected their resources appropriately. we are not looking to create a generic backup solution for your whole cluster: 95% of our users will probably want to just create a single scheduled backup and do nothing special. In which case having extra resources will just make the configuration more complex. If we were to go down a more velero-like approach, i'd rather work on building an integration to velero so people can just use that instead.

Reply:
I'm fine not worrying about this for now and just sticking with `gcsBucket`. We can always change it later.